src/resolve-dependency.ts

```diff
@@ -1,28 +1,34 @@
-const { isAbsolute, resolve, sep } = require('path');
+import { isAbsolute, resolve, sep } from 'path';
+import { Job } from './node-file-trace';
 
 // node resolver
 // custom implementation to emit only needed package.json files for resolver
 // (package.json files are emitted as they are hit)
-module.exports = function resolveDependency (specifier, parent, job, cjsResolve = true) {
-  let resolved;
+export default function resolveDependency (specifier: string, parent: string, job: Job, cjsResolve = true) {
+  let resolved: string | string[] | undefined;
   if (isAbsolute(specifier) || specifier === '.' || specifier === '..' || specifier.startsWith('./') || specifier.startsWith('../')) {
     const trailingSlash = specifier.endsWith('/');
     resolved = resolvePath(resolve(parent, '..', specifier) + (trailingSlash ? '/' : ''), parent, job);
-  }
-  else {
+  } else {
     resolved = resolvePackage(specifier, parent, job, cjsResolve);
   }
-  if (typeof resolved === 'string' && resolved.startsWith('node:')) return resolved;
-  if (Array.isArray(resolved))
+  if (typeof resolved === 'string' && !resolved.startsWith('node:')) {
+    return job.realpath(resolved, parent);
+  } else if (Array.isArray(resolved)) {
     return resolved.map(resolved => job.realpath(resolved, parent));
-  return job.realpath(resolved, parent);
+  }
+  return resolved;
```

Comment:
Is it expected that `resolved` can be undefined here?

Reply:
`resolved` will never be undefined here, and the typing should be able to capture that?

Reply:
There's two calls above, one to `resolvePath()` and another to `resolvePackage()` and both can return undefined. What makes you think that `resolved` will never be `undefined` here?

Reply:
Both `resolvePath` and `resolvePackage` should never return `undefined`.

Reply:
@guybedford But `resolvePath()` calls `resolveFile()` and [you mentioned](https://github.com/vercel/node-file-trace/pull/133#discussion_r453301867) that `resolveFile()` can return undefined.

Therefore, `resolvePath()` can return `undefined` and this `undefined` case is not handled in `resolveDependency()` so `resolveDependency()` can returned `undefined` as well.

Perhaps typescript is revealing a case we didn't think about and it should be handled?

Reply:
`resolvePath` will always explicitly throw an error for no resolution, and can't return `undefined`, while `resolveFile` can.

Reply:
@guybedford What about `resolvePackage()`? It returns `resolveFile()` which could be undefined.

That still means that `resolveDependency()` can return undefined as well.

Reply:
No, `resolvePackage` should always return a result or throw. If there are any paths that don't do that that is a bug - I've just commented on one possible case.

Reply:
Looks good now after https://github.com/vercel/node-file-trace/pull/133#discussion_r453679297 👍